File: src/main/java/com/itsci/mju/maebanjumpen/service/impl/HireServiceImpl.java
package com.itsci.mju.maebanjumpen.service.impl;

import com.itsci.mju.maebanjumpen.dto.HireDTO;
import com.itsci.mju.maebanjumpen.exception.HirerNotFoundException;
import com.itsci.mju.maebanjumpen.exception.HousekeeperNotFoundException;
import com.itsci.mju.maebanjumpen.exception.InsufficientBalanceException;
import com.itsci.mju.maebanjumpen.mapper.HireMapper;
import com.itsci.mju.maebanjumpen.model.Hire;
import com.itsci.mju.maebanjumpen.model.Hirer;
import com.itsci.mju.maebanjumpen.model.Housekeeper;
import com.itsci.mju.maebanjumpen.model.SkillType;
import com.itsci.mju.maebanjumpen.repository.HireRepository;
import com.itsci.mju.maebanjumpen.repository.HirerRepository;
import com.itsci.mju.maebanjumpen.repository.HousekeeperRepository;
import com.itsci.mju.maebanjumpen.repository.SkillTypeRepository;
import com.itsci.mju.maebanjumpen.service.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;

@Service
@RequiredArgsConstructor
public class HireServiceImpl implements HireService {

    private final HireMapper hireMapper;
    private final HireRepository hireRepository;
    private final HirerService hirerService;
    private final HousekeeperService housekeeperService;
    private final HousekeeperSkillService housekeeperSkillService;
    private final SkillTypeRepository skillTypeRepository;
    private final HirerRepository hirerRepository;
    private final HousekeeperRepository housekeeperRepository;
    private final HireStatusUpdateService statusUpdateService;

    @Override
    @Transactional(readOnly = true)
    public List<HireDTO> getAllHires() {
        List<Hire> entities = hireRepository.findAllWithDetails(); // 💡 ใช้ findAllWithDetails เพื่อโหลดรายละเอียด
        return hireMapper.toDtoList(entities);
    }

    @Override
    @Transactional(readOnly = true)
    public HireDTO getHireById(Integer id) {
        Hire hire = hireRepository.fetchByIdWithAllDetails(id)
                .orElseThrow(() -> new IllegalArgumentException("Hire with ID " + id + " not found."));
        return hireMapper.toDto(hire);
    }

    @Override
    @Transactional(readOnly = true)
    public List<HireDTO> getHiresByHirerId(Integer hirerId) {
        List<Hire> hires = hireRepository.findByHirerIdWithDetails(hirerId);
        return hireMapper.toDtoList(hires);
    }

    @Override
    @Transactional(readOnly = true)
    public List<HireDTO> getHiresByHousekeeperId(Integer housekeeperId) {
        List<Hire> hires = hireRepository.findByHousekeeperIdWithDetails(housekeeperId);
        return hireMapper.toDtoList(hires);
    }

    /**
     * 💡 NEW SERVICE METHOD: ดึงเฉพาะงานจ้างที่เสร็จสมบูรณ์ ('Completed') สำหรับ Housekeeper ID นั้นๆ
     * ใช้สำหรับคำนวณ 'Jobs Done' และดึง Reviews
     */
    @Override
    @Transactional(readOnly = true)
    public List<HireDTO> getCompletedHiresByHousekeeperId(Integer housekeeperId) {
        final String COMPLETED_STATUS = "Completed";
        List<Hire> completedHires = hireRepository.findByHousekeeperIdAndJobStatusWithDetails(housekeeperId, COMPLETED_STATUS);
        return hireMapper.toDtoList(completedHires);
    }

    @Override
    @Transactional
    public HireDTO saveHire(HireDTO hireDto) {
        Hire hire = hireMapper.toEntity(hireDto);

        Hirer hirer = validateAndGetHirer(hire);
        hire.setHirer(hirer);

        Housekeeper housekeeper = validateAndGetHousekeeper(hire);
        hire.setHousekeeper(housekeeper);

        SkillType skillType = validateAndGetSkillType(hire);
        hire.setSkillType(skillType);

        if (hireDto.getPaymentAmount() == null || hireDto.getPaymentAmount() <= 0) {
            throw new IllegalArgumentException("Payment amount is required and must be a positive value.");
        }
        hire.setPaymentAmount(hireDto.getPaymentAmount());

        if (hirer.getBalance() < hire.getPaymentAmount()) {
            throw new InsufficientBalanceException("Insufficient balance to create a hire.");
        }

        hire.setHireName(skillType.getSkillTypeName());
        hire.setHireDetail(hireDto.getHireDetail());
        Hire savedHire = hireRepository.save(hire);
        Hire finalHire = hireRepository.fetchByIdWithAllDetails(savedHire.getHireId()).orElse(savedHire);
        return hireMapper.toDto(finalHire);
    }

    @Override
    @Transactional
    public HireDTO updateHire(Integer id, HireDTO hireDto)
            throws InsufficientBalanceException, HirerNotFoundException {

        Hire existingHire = hireRepository.fetchByIdWithAllDetails(id)
                .orElseThrow(() -> new IllegalArgumentException("Hire with ID " + id + " not found."));

        String oldStatus = existingHire.getJobStatus();
        String newStatus = hireDto.getJobStatus();

        if (newStatus != null && "Completed".equalsIgnoreCase(newStatus)
                && !"Completed".equalsIgnoreCase(oldStatus)) {

            if (existingHire.getPaymentAmount() == null || existingHire.getPaymentAmount() <= 0) {
                throw new IllegalStateException("Cannot complete hire. Payment amount is missing or invalid.");
            }

            hirerService.deductBalance(existingHire.getHirer().getId(), existingHire.getPaymentAmount());
            housekeeperService.addBalance(existingHire.getHousekeeper().getId(), existingHire.getPaymentAmount());

            housekeeperSkillService.updateSkillLevelAndHiresCompleted(
                    existingHire.getHousekeeper().getId(),
                    existingHire.getSkillType().getSkillTypeId()
            );
        }

        if (newStatus != null) existingHire.setJobStatus(newStatus);
        if (hireDto.getHireName() != null) existingHire.setHireName(hireDto.getHireName());
        if (hireDto.getHireDetail() != null) existingHire.setHireDetail(hireDto.getHireDetail());
        if (hireDto.getHireDate() != null) existingHire.setHireDate(hireDto.getHireDate());
        if (hireDto.getStartDate() != null) existingHire.setStartDate(hireDto.getStartDate());
        if (hireDto.getStartTime() != null) existingHire.setStartTime(hireDto.getStartTime());
        if (hireDto.getEndTime() != null) existingHire.setEndTime(hireDto.getEndTime());
        if (hireDto.getLocation() != null) existingHire.setLocation(hireDto.getLocation());

        if (hireDto.getSkillType() != null
                && hireDto.getSkillType().getSkillTypeId() != null
                && !existingHire.getSkillType().getSkillTypeId()
                .equals(hireDto.getSkillType().getSkillTypeId())) {
            SkillType newSkillType = skillTypeRepository.findById(
                    hireDto.getSkillType().getSkillTypeId()
            ).orElseThrow(() ->
                    new IllegalArgumentException("SkillType with ID "
                            + hireDto.getSkillType().getSkillTypeId() + " not found."));
            existingHire.setSkillType(newSkillType);
        }

        Hire updatedHire = hireRepository.save(existingHire);
        Hire finalHire = hireRepository.fetchByIdWithAllDetails(updatedHire.getHireId()).orElse(updatedHire);
        return hireMapper.toDto(finalHire);
    }

    @Override
    @Transactional
    public void deleteHire(Integer id) {
        hireRepository.deleteById(id);
    }

    @Override
    @Transactional
    public HireDTO addProgressionImagesToHire(Integer hireId, List<String> imageUrls) {
        Hire hire = hireRepository.fetchByIdWithAllDetails(hireId)
                .orElseThrow(() -> new IllegalArgumentException("Hire with ID " + hireId + " not found."));
        if (hire.getProgressionImageUrls() == null) hire.setProgressionImageUrls(new ArrayList<>());
        hire.getProgressionImageUrls().addAll(imageUrls);

        Hire updatedHire = hireRepository.save(hire);
        return hireMapper.toDto(updatedHire);
    }

    private Hirer validateAndGetHirer(Hire hire) {
        if (hire.getHirer() == null || hire.getHirer().getId() == null) {
            throw new IllegalArgumentException("Hirer ID is required for creating a hire.");
        }
        return hirerRepository.findById(hire.getHirer().getId())
                .orElseThrow(() -> new HirerNotFoundException("Hirer with ID "
                        + hire.getHirer().getId() + " not found."));
    }

    private Housekeeper validateAndGetHousekeeper(Hire hire) {
        if (hire.getHousekeeper() == null || hire.getHousekeeper().getId() == null) {
            throw new IllegalArgumentException("Housekeeper ID is required for creating a hire.");
        }
        return housekeeperRepository.findById(hire.getHousekeeper().getId())
                .orElseThrow(() -> new HousekeeperNotFoundException("Housekeeper with ID "
                        + hire.getHousekeeper().getId() + " not found."));
    }

    private SkillType validateAndGetSkillType(Hire hire) {
        if (hire.getSkillType() == null || hire.getSkillType().getSkillTypeId() == null) {
            throw new IllegalArgumentException("SkillType ID is required for creating a hire.");
        }
        return skillTypeRepository.findById(hire.getSkillType().getSkillTypeId())
                .orElseThrow(() -> new IllegalArgumentException("SkillType with ID "
                        + hire.getSkillType().getSkillTypeId() + " not found."));
    }
}

File: src/main/kotlin/com/itsci/mju/maebanjumpen/hire/service/impl/HireServiceImpl.kt
package com.itsci.mju.maebanjumpen.hire.service.impl

import com.itsci.mju.maebanjumpen.entity.Hire
import com.itsci.mju.maebanjumpen.entity.Hirer
import com.itsci.mju.maebanjumpen.entity.Housekeeper
import com.itsci.mju.maebanjumpen.entity.SkillType
import com.itsci.mju.maebanjumpen.exception.HirerNotFoundException
import com.itsci.mju.maebanjumpen.exception.HousekeeperNotFoundException
import com.itsci.mju.maebanjumpen.exception.InsufficientBalanceException
import com.itsci.mju.maebanjumpen.hire.dto.HireDTO
import com.itsci.mju.maebanjumpen.hire.repository.HireRepository
import com.itsci.mju.maebanjumpen.hire.service.HireService
import com.itsci.mju.maebanjumpen.housekeeperskill.service.HousekeeperSkillService
import com.itsci.mju.maebanjumpen.mapper.HireMapper
import com.itsci.mju.maebanjumpen.partyrole.repository.HirerRepository
import com.itsci.mju.maebanjumpen.partyrole.repository.HousekeeperRepository
import com.itsci.mju.maebanjumpen.partyrole.service.HirerService
import com.itsci.mju.maebanjumpen.partyrole.service.HousekeeperService
import com.itsci.mju.maebanjumpen.skilltype.repository.SkillTypeRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class HireServiceImpl(
    private val hireMapper: HireMapper,
    private val hireRepository: HireRepository,
    private val hirerService: HirerService,
    private val housekeeperService: HousekeeperService,
    private val housekeeperSkillService: HousekeeperSkillService,
    private val skillTypeRepository: SkillTypeRepository,
    private val hirerRepository: HirerRepository,
    private val housekeeperRepository: HousekeeperRepository,
    private val statusUpdateService: HireStatusUpdateService
) : HireService {

    @Transactional(readOnly = true)
    override fun getAllHires(): List<HireDTO> {
        val entities = hireRepository.findAllWithDetails()
        return hireMapper.toDtoList(entities)
    }

    @Transactional(readOnly = true)
    override fun getHireById(id: Int): HireDTO {
        val hire = hireRepository.fetchByIdWithAllDetails(id)
            .orElseThrow { IllegalArgumentException("Hire with ID $id not found.") }
        return hireMapper.toDto(hire)
    }

    @Transactional(readOnly = true)
    override fun getHiresByHirerId(hirerId: Int): List<HireDTO> {
        val hires = hireRepository.findByHirerIdWithDetails(hirerId)
        return hireMapper.toDtoList(hires)
    }

    @Transactional(readOnly = true)
    override fun getHiresByHousekeeperId(housekeeperId: Int): List<HireDTO> {
        val hires = hireRepository.findByHousekeeperIdWithDetails(housekeeperId)
        return hireMapper.toDtoList(hires)
    }

    @Transactional(readOnly = true)
    override fun getCompletedHiresByHousekeeperId(housekeeperId: Int): List<HireDTO> {
        val completedStatus = "Completed"
        val completedHires = hireRepository.findByHousekeeperIdAndJobStatusWithDetails(housekeeperId, completedStatus)
        return hireMapper.toDtoList(completedHires)
    }

    @Transactional
    override fun saveHire(hireDto: HireDTO): HireDTO {
        val hire = hireMapper.toEntity(hireDto)

        val hirer = validateAndGetHirer(hire)
        hire.hirer = hirer

        val housekeeper = validateAndGetHousekeeper(hire)
        hire.housekeeper = housekeeper

        val skillType = validateAndGetSkillType(hire)
        hire.skillType = skillType

        if (hireDto.paymentAmount == null || hireDto.paymentAmount!! <= 0) {
            throw IllegalArgumentException("Payment amount is required and must be a positive value.")
        }
        hire.paymentAmount = hireDto.paymentAmount!!

        if ((hirer.balance ?: 0.0) < hire.paymentAmount) {
            throw InsufficientBalanceException("Insufficient balance to create a hire.")
        }

        hire.hireName = skillType.skillTypeName ?: ""
        hire.hireDetail = hireDto.hireDetail ?: ""
        val savedHire = hireRepository.save(hire)
        val finalHire = hireRepository.fetchByIdWithAllDetails(savedHire.hireId!!).orElse(savedHire)
        return hireMapper.toDto(finalHire)
    }

    @Transactional
    override fun updateHire(id: Int, hireDto: HireDTO): HireDTO {
        val existingHire = hireRepository.fetchByIdWithAllDetails(id)
            .orElseThrow { IllegalArgumentException("Hire with ID $id not found.") }

        val oldStatus = existingHire.jobStatus
        val newStatus = hireDto.jobStatus

        if (newStatus != null && newStatus.equals("Completed", ignoreCase = true)
            && !oldStatus.equals("Completed", ignoreCase = true)) {

            if (existingHire.paymentAmount <= 0) {
                throw IllegalStateException("Cannot complete hire. Payment amount is missing or invalid.")
            }

            hirerService.deductBalance(existingHire.hirer!!.id!!, existingHire.paymentAmount)
            housekeeperService.addBalance(existingHire.housekeeper!!.id!!, existingHire.paymentAmount)

            housekeeperSkillService.updateSkillLevelAndHiresCompleted(
                existingHire.housekeeper!!.id!!,
                existingHire.skillType!!.skillTypeId!!
            )
        }

        newStatus?.let { existingHire.jobStatus = it }
        hireDto.hireName?.let { existingHire.hireName = it }
        hireDto.hireDetail?.let { existingHire.hireDetail = it }
        hireDto.hireDate?.let { existingHire.hireDate = it }
        hireDto.startDate?.let { existingHire.startDate = it }
        hireDto.startTime?.let { existingHire.startTime = it }
        hireDto.endTime?.let { existingHire.endTime = it }
        hireDto.location?.let { existingHire.location = it }

        if (hireDto.skillType?.skillTypeId != null
            && existingHire.skillType?.skillTypeId != hireDto.skillType?.skillTypeId) {
            val newSkillType = skillTypeRepository.findById(hireDto.skillType!!.skillTypeId!!)
                .orElseThrow { IllegalArgumentException("SkillType with ID ${hireDto.skillType?.skillTypeId} not found.") }
            existingHire.skillType = newSkillType
        }

        val updatedHire = hireRepository.save(existingHire)
        val finalHire = hireRepository.fetchByIdWithAllDetails(updatedHire.hireId!!).orElse(updatedHire)
        return hireMapper.toDto(finalHire)
    }

    @Transactional
    override fun deleteHire(id: Int) {
        hireRepository.deleteById(id)
    }

    @Transactional
    override fun addProgressionImagesToHire(hireId: Int, imageUrls: List<String>): HireDTO {
        val hire = hireRepository.fetchByIdWithAllDetails(hireId)
            .orElseThrow { IllegalArgumentException("Hire with ID $hireId not found.") }
        hire.progressionImageUrls.addAll(imageUrls)

        val updatedHire = hireRepository.save(hire)
        return hireMapper.toDto(updatedHire)
    }

    private fun validateAndGetHirer(hire: Hire): Hirer {
        if (hire.hirer?.id == null) {
            throw IllegalArgumentException("Hirer ID is required for creating a hire.")
        }
        return hirerRepository.findById(hire.hirer!!.id!!)
            .orElseThrow { HirerNotFoundException("Hirer with ID ${hire.hirer?.id} not found.") }
    }

    private fun validateAndGetHousekeeper(hire: Hire): Housekeeper {
        if (hire.housekeeper?.id == null) {
            throw IllegalArgumentException("Housekeeper ID is required for creating a hire.")
        }
        return housekeeperRepository.findById(hire.housekeeper!!.id!!)
            .orElseThrow { HousekeeperNotFoundException("Housekeeper with ID ${hire.housekeeper?.id} not found.") }
    }

    private fun validateAndGetSkillType(hire: Hire): SkillType {
        if (hire.skillType?.skillTypeId == null) {
            throw IllegalArgumentException("SkillType ID is required for creating a hire.")
        }
        return skillTypeRepository.findById(hire.skillType!!.skillTypeId!!)
            .orElseThrow { IllegalArgumentException("SkillType with ID ${hire.skillType?.skillTypeId} not found.") }
    }
}

